Feature: repeatable

ayush2k opened the pull request:
`DCO 1.1 Signed-off-by: {{Ayush Gupta}} <{{ayush2k@gmail.com}}>`

ayush2k commented:
@scottnath @Snugug

scottnath commented on lib/content-types/only.js:
```diff
@@ -16,16 +16,32 @@ const only = (type, config) => {
-        attr.inputs[input].name = attribute.inputs[input].name;
-      });
+      // Flattens multiple instances
+      if (Array.isArray(config[attr.id])) {
```
For both multi-selects and checkboxes, they will be an array.

But, selects will not have `--index`

ayush2k replied:
Inputs would be an array for the multiple instances. 
The value of select and checkbox will be stored as string and array respectively

scottnath commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+
+          // Add error messaging and required
+          render += addRequired(addError(html, input, errors), input);
+          render += `<input type="button" value="delete" class="delete-button" id="${input.id}-del--${index}">`;
```
make these `<button>` elements instead

scottnath commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+          render += `<input type="button" value="delete" class="delete-button" id="${input.id}-del--${index}">`;
+          render += '</div>';
+        });
+        render += `<input type="button" value="add" class="add-button" id="${input.id}-add">`;
```
`<button>`

note: added a [new issue for having the `add` text i18 compliant](https://github.com/punchcard-cms/content-types/issues/54) - does not need to be in this PR.

scottnath commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,248 @@
+            }
+          }
+          else if (child.htmlFor && child.htmlFor != "") {
+            // htmlFor = child.htmlFor.split('--');
```
what are these commented lines here for?

scottnath commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,248 @@
+            child.id = newId;
+          }
+
+          if (child.type == 'text') {
```
There could any number of `type`s other than `text` (url, email, etc). This check should be done in a different way.

scottnath commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,248 @@
+    };
+
+    // Adds event listener for existing delete button
+    document.querySelectorAll('input.delete-button').forEach(function (button) {
```
when you change to `<button` this will need to be changed too

scottnath commented on tests/content-types.js:
```diff
@@ -3,20 +3,302 @@ import config from 'config';
-    'my-textarea': {
-      textarea: {
-        value: 'Hello World I am here'
+  return only({
```
Please pull out these fixtures and put them in a file in the fixtures directory

Snugug commented on lib/content-types/only.js:
```diff
@@ -16,16 +16,32 @@ const only = (type, config) => {
+      // Flattens multiple instances
+      if (Array.isArray(config[attr.id])) {
+        const instance = [];
+        config[attr.id].forEach((data, index) => {
```
Can you write this as a `.map` instead of a `forEach`? That'll simplify this code a lot and remove your need to add the index explicitly

ayush2k replied:
done. I still used `index` for indexing ids of element.

Snugug commented on lib/form/html.js:
```diff
@@ -113,10 +113,14 @@ const addRequired = (html, input) => {
     let rendered = type.attributes.map(input => {
-      const inputs = Object.keys(input.inputs).length;
+      let inputs = Object.keys(input.inputs).length;
+      if (Array.isArray(input.inputs) && input.inputs.length > 0) {
```
Can you explain to me what you're doing here? Why are hard-coding the first item to grab?

ayush2k replied:
Earlier we were using the length of inputs, which was an object. 
In the case of repeatables, `Object.keys(input.inputs).length`, will return the length of array instead.
In order to get the length of inputs in input-plugin, I hard-coded to get the length of inputs.

Snugug replied:
mk

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+      // Wraps repeatable attributes around div
+      if (Array.isArray(input.inputs)) {
+        context.forEach((data, index) => {
+          render += `<div id="sub-${input.id}" class="form-repeatable">`;
```
Two dashes here, `form--repeatable`

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+
+          // Add error messaging and required
+          render += addRequired(addError(html, input, errors), input);
+          render += `<button type="button" class="delete-button" id="${input.id}-del--${index}">Delete</button>`;
```
`form--delete` for the class please

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+          render += `<button type="button" class="delete-button" id="${input.id}-del--${index}">Delete</button>`;
+          render += '</div>';
+        });
+        render += `<button type="button" class="add-button" id="${input.id}-add">Add</button>`;
```
`form--add` for the class please

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+        render += `<button type="button" class="add-button" id="${input.id}-add">Add</button>`;
+      }
+      else if (input.repeatable && input.repeatable === true) {
+        render += `<div id="sub-${input.id}" class="form-repeatable">`;
```
`form--repeatable`

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+        // Add error messaging and required
+        render += addRequired(addError(html, input, errors), input);
+        render += '</div>';
+        render += `<button type="button" class="add-button" id="${input.id}-add">Add</button>`;
```
`form--add`

Snugug replied:
Why are you rendering the button outside of the containing div? Wouldn't we want this wrapped to allow us to style them together?

ayush2k replied:
`<div id="sub-${input.id}" class="form-repeatable">`, contains the input plugin which is replicated.
If I keep add-button inside this div, then we would have multiple add for a single attribute.
I assume add button can be styled using  `<div id="${input.id}" class="form--field ${required}">`

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,245 @@
+  // Iterates through every instance of input-plugins
+  Object.keys(allSettings).forEach(function (input) {
+    var idAdd = input + '-add';
+    var idDel = input + '-del';
```
Please use the full word `delete` here and anywhere else this is referenced

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,245 @@
+      if (allRepeatables[input].length == 1) {
+        var newElement = document.createElement('button');
+        newElement.type = 'button';
+        newElement.innerHTML = 'Delete';
```
Please do not use `innerHTML`, it's creates the potential for security vulnerabilities. Please use `textContent` instead

Snugug commented:
I see lots of mixing of single and double quotes in the browser JS. Can you please make sure that's being run through linting?

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,245 @@
+
+      // Increments length of attribute
+      allRepeatables[input].length++;
+      var containerSet = container.parentNode.querySelectorAll('div.form-repeatable');
```
Is there an instance where `.form--repeatable` isn't on a div? We don't need to over-qualify this selector

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,245 @@
+    };
+
+    // Adds event listener for existing delete button
+    document.querySelectorAll('button.delete-button').forEach(function (button) {
```
Just need the class, not the tag as well

Snugug commented on lib/form/js/validation.js:
```diff
@@ -2,7 +2,7 @@
 
 // TODO: Add In-Browser Tests
-(function formValidation() {
+var validate = (function formValidation() {
```
Instead of setting the whole function to a variable, export the function from the closure by setting `window.validate` to the function you want to export

Snugug commented on tests/content-types.js:
```diff
@@ -3,20 +3,38 @@ import config from 'config';
 import util from 'util';
 import fs from 'fs';
+import only from '../lib/content-types/only.js';
+var bar_input = require('./fixtures/objects//bar_input.js')
```
Please use the same import statements as above. This is something linting should have caught and didn't. 

I'm going to need to go in and make sure our linter is working correctly before we can merge this as I'm seeing lots of linting issues in your code

Snugug commented on tests/content-types.js:
```diff
@@ -3,20 +3,38 @@ import config from 'config';
-    'my-textarea': {
-      textarea: {
-        value: 'Hello World I am here'
+  return only(bar_input, {
```
@snugug start here after linting

Snugug commented:
@ayush2k Content Types updated with linting in place. Please update your branch (there may be conflicts) before we continue

ayush2k commented:
@Snugug @scottnath 
Updated

Snugug commented on tests/content-types.js:
```diff
@@ -1,16 +1,33 @@
 import test from 'ava';
 import types from '../lib/content-types';
+import only from '../lib/content-types/only.js';
+import barInput from './fixtures/objects//bar_input.js';
```
This has two dashes in its url

Snugug commented on tests/fixtures/objects/bar_expected.js:
```diff
@@ -0,0 +1,187 @@
+        id: 'something-new',
+        inputs: {
+          text: {
+            id: '5ba58361-87b7-4e79-8ea0-e54635ad23bb',
```
Is this ID really reliable to use? These are all randomly generated

Snugug replied:
nm

Snugug commented on tests/form.js:
```diff
@@ -2,6 +2,7 @@ import test from 'ava';
 import types from '../lib/content-types';
 import form from '../lib/form';
 import includes from 'lodash/includes';
+import barInput from './fixtures/objects/bar_expected.js';
```
File names should be `kebab-case` not `snake_case`

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+
+          // Add error messaging and required
+          render += addRequired(addError(html, input, errors), input);
+          render += `<button type="button" class="delete--button" id="${input.id}--delete--${index}">Delete</button>`;
```
`form--delete` not `delete--button`

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+          render += `<button type="button" class="delete--button" id="${input.id}--delete--${index}">Delete</button>`;
+          render += '</div>';
+        });
+        render += `<button type="button" class="add--button" id="${input.id}--add">Add</button>`;
```
`form--add` not `add--button`

Snugug commented on lib/form/html.js:
```diff
@@ -137,8 +141,31 @@ const renderer = (type, errors) => {
+      // Wraps repeatable attributes around div
+      if (Array.isArray(input.inputs)) {
+        context.forEach((data, index) => {
+          render += `<div id="sub-${input.id}" class="form--repeatable">`;
```
What is `sub`

ayush2k replied:
it creates `sub div` for each instance

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,266 @@
+        addHandler,
+        addButton;
+
+    deleteHandler = function deleteEventHandler(e) {
```
You appear to be creating this function for each and every setting. Is that true? We really don't want to be creating functions inside of loops; can you have these functions built separately outside of the loop and just loop over the settings to apply the functions as needed?

Snugug replied:
Where does this function end? I can't seem to follow the parenthesis correctly.

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,266 @@
+      parent.removeChild(node);
+      allRepeatables[input].length--;
+
+      deleteNode = function deleteElement(key) {
```
It _appears_ as if this function is created inside the `deleteHandler` function, and I'm fairly confident that's not your intention. Please make sure all of your parenthesis are in place

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,266 @@
+              delete allIDs[oldChildId];
+            }
+          }
+          childName = child.getAttribute('name');
```
Either do all of the assignments at the top or have comments and space between new logical sections please

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,266 @@
+      };
+
+      // change ids and names of subsequest instances
+      while (siblingNode) {
```
Please group function definitions together before writing the code external to the function that consumes said functions

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,266 @@
+      }
+
+      if (allRepeatables[input].length === 1) {
+        container = parent.querySelectorAll('.form--repeatable')[0];
```
If you always want to grab the first item, `document.querySelector` should be used instead

Snugug replied:
Applies to all places this is happening

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,266 @@
+      button.addEventListener('click', deleteHandler);
+    });
+
+    addButton = document.getElementById(idAdd);
```
This looks like there will only be one add button?

Snugug commented:
Each of the functions in `form/js/repeatable.js` seem to be doing a lot of work. I'd like to see each logical grouping of work its doing broken out in to functions with a larger function to bring them all together. It'll be easier to maintain having these functions be smaller and more targeted.

I'd also recommend looking in to the array functions, `map`, `filter`, `reduce`. They are chain-able and will very likely reduce the amount of `forEach` and `if` statements you have to deal with in your code, cleaning it up a bunch.

Snugug commented:
@ayush2k I've got some comments for you. If you could look them over and resolve them, we should be good to merge

ayush2k commented:
@Snugug I have been working on them. 
Should have them committed in 30-45 minutes.

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -4,257 +4,338 @@
 (function formRepeatable() {
   'use strict';
 
+  // Disabled one-var and func-names for better readability
```
Please do not disable them, they are there for a reason.

Snugug replied:
`func-names`, for instance, aids in debugging when functions exist within closures (like they do here) and `one-var` ensures we know what's going to be variable at the top of our function scope so there are no surprises

ayush2k replied:
Due to func-names, we will have two names for a same function
Do you want me to add another name to all functions?

Snugug replied:
Yes. That is exactly why this rule exists.
http://eslint.org/docs/rules/func-names

ayush2k replied:
Now I get why we need to name them!

ayush2k replied:
These are disabled in validation.js as well. I will enable and structure them accordingly.

ayush2k commented:
@Snugug resolved and enabled linter warning. 
Please review
Thanks!

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -4,257 +4,338 @@
+    container.id = container.id + '--' + (index - 1);
+
+    // Updates all children of container
+    container.querySelectorAll('*').forEach(function setAttributes(node) {
```
Did you test this across browsers? [`querySelectorAll`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll) returns a [`NodeList`](https://developer.mozilla.org/en-US/docs/Web/API/NodeList) which is _not_ an Array and does not have `forEach` available. In browsers that implement `NodeList` properly, this will break.

You should either loop over these with a `for` loop, or use `Array.prototype.forEach.call()` instead

Snugug replied:
This holds true for all places you do this

Snugug commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,346 @@
+        container.id = container.id + '--' + (index - 1);
+
+        // Updates all children of container
+        container.querySelectorAll('*').forEach(function setAttributes(node) {
```
Did you test this across browsers? [`querySelectorAll`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll) returns a [`NodeList`](https://developer.mozilla.org/en-US/docs/Web/API/NodeList) which is _not_ an Array and does not have `forEach` available. In browsers that implement `NodeList` properly, this will break.

You should either loop over these with a `for` loop, or use `Array.prototype.forEach.call()` instead. This holds true for all places you do this

ayush2k replied:
That is true, forEach doesn't work for nodeList.
Will test it in other browsers and use Array.prototype.forEach.call() instead

Snugug commented:
@ayush2k One more comment, on cross-browser compatibility of some of your code. Take a look and, once updated, just about there. Would like @scottnath to have a once-over as well

Snugug commented on lib/form/js/validation.js:
```diff
@@ -6,90 +6,91 @@
-        // Remove invalid
-        target.removeAttribute('aria-invalid');
+        // Set all sibling input and settings
+        Object.keys(inp.siblings).forEach(function validateSiblings(sibling) {
```
This breaks with repeatables

Snugug commented:
* Your `Form Generation, with repeatables` test, while it passes, the actual HTML and JS don't work, so I'm not sure if that's a valid test
* Please update in `package.json` line 15 to be `"lint:node": "eslint -c ./.eslintrc-node.yml index.js './lib/*.js' './lib/content-types/**/*.js' './lib/form/*.js' './tests/fixtures/**/*.js' ",` and re-run your linting; there are a bunch of linting errors in the fixtures.
* Testing this in browser with the `bar` content type, repeatables break the validation script as the validation script doesn't know that you've changed the IDs on it. It's still looking for siblings with the same non-repeatable ID, but they all have repeatable IDs now so when it goes to deal with that, it can't find the right siblings. Please update validation and ensure that it works with repeatables.

Snugug commented:
* In the `bar` content type (which I am testing with), `email` is suppose to be repeatable with a minimum of 1 and maximum of 3 times. In the rendered HTML and JS, `email` is not repeatable. Please make sure that it is, and make sure that validation is not broken on it when made available

ayush2k commented:
@Snugug 

- updated sibling for the validation. I hope it works now. Tested it with quote and checkboxes.
- Currently, the program is considering repeatable as boolean. I will start making other changes to making it work for min and max.

ayush2k commented on lib/form/js/repeatable.js:
```diff
@@ -0,0 +1,396 @@
+        element.setAttribute('class', 'form--add');
+
+        // rule disable as adding 'add' button is required in deleteHandler and addHandler requires to add 'Delete' button
+        element.addEventListener('click', addHandler); // eslint-disable-line no-use-before-define
```
@scottnath:

ayush2k commented:
@Snugug @scottnath 
Please review.
- Added functionality for min and max values
- Fixed validation and errors 
- Disabled a linter warning because circular dependencies of functions. Please suggest how can I fix it.

scottnath commented:
@Snugug @ayush2k I don't see any issues here. Deferring to Sam for final say.

Snugug commented:
This all looks good. There is are just _two_ more (hopefully fairly small) thing that needs to be done for this to merge:

* When doing the first-time render of an HTML form, the number of repeatable elements that should be rendered should be equal to the `min` value of the repeatable. Likewise, being able to delete fields should stop when the number of available fields equals the `min` number.
* When initially rendering a repeatable field, it's `name`should always include its index. Features like [autofill](https://cloudfour.com/thinks/autofill-what-web-devs-should-know-but-dont/) rely on the input `name`, so having the name change out from underneath it will break this functionality.

Thanks! If you can get these two things in and tests pass Monday, this will be merged Monday!

Snugug commented on lib/content-types/merge.js:
```diff
@@ -101,6 +101,31 @@ const squish = (types) => {
         }
 
+        // Organizes repeatable attributes
+        if (attribute.repeatable) {
```
`attribute.hasOwnProperty('repeatable')`

Snugug commented on lib/content-types/merge.js:
```diff
@@ -101,6 +101,31 @@ const squish = (types) => {
+        if (attribute.repeatable) {
+          if (typeof attribute.repeatable === 'object') {
+            plugin.repeatable = {};
+            if (!attribute.repeatable.hasOwnProperty('min')) {
```
Some comments between these if statements to describe what they do, please

ayush2k commented:
@Snugug 
please review.

Snugug commented:
@ayush2k The initial rendering of the multiple repeatables needs to be done server-side, not client-side. Other than that, looking good. Going to do some cross-browser testing

Snugug commented:
Now that I'm really digging in to the functionality, I'm noticing some a quirk: if a field has an error, and that field is deleted, the error is not, but should be

Snugug commented on lib/form/scripts.js:
```diff
@@ -26,6 +26,16 @@ const rendered = (type) => {
     });
     const existingPlugins = {};
+
+    // Checks if plugin directory has values
```
You have this check in here too. What's this doing here?

ayush2k commented:
@Snugug Updated this PR to latest.

Snugug commented on lib/content-types/merge.js:
```diff
@@ -91,32 +91,26 @@ const squish = (types) => {
         plugin = _.cloneDeep(plugins[`input-plugin-${attribute.type}`]);
 
-        if (attribute.hasOwnProperty('name')) {
+        if (attribute.name) {
```
What happened here? Why was this reverted? What happened to the comments?

Snugug commented on lib/form/js/validation.js:
```diff
@@ -6,90 +6,91 @@
+        var target = e.target,
+            value = target.value,
+            inp = allInputs[target.id],
+            parent = target.closest('div.form--repeatable, div.form--field, fieldset.form--fieldset'),
```
Have you tested this cross-browser? It [isn't implemented in IE](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest#Browser_compatibility)

Snugug replied:
And these selectors are over-qualified with the `div`s, should only be the class

Snugug commented:
@ayush2k The only outstanding issue is the `closest`. Once that's resolved, I'll look again, but close to merging

ayush2k commented:
@Snugug fixed closest issue. Now looping through the parent nodes

ayush2k commented:
@Snugug 
when testing with safari, I got this error for the input-plugin [line](https://github.com/punchcard-cms/input-plugin-checkbox/blob/master/index.js#L9)
and probably will get this for all other input-plugins.

`SyntaxError: Unexpected keyword 'const'. Const declarations are not supported in strict mode.`

Snugug commented on lib/form/js/validation.js:
```diff
@@ -6,90 +6,100 @@
-    settings.target = inp.settings;
-    settings.all = {};
-    settings.all[inp.name] = inp.settings;
+        while (parent.className.indexOf('form--repeatable') === -1 &&
```
A couple of thoughts on this:

* Instead of using `className.indexOf` (which would be `true` for something like `bar-form--repeatable` which is not what we want), use `classList.contains('form--repeatable')` as that will tell you if the actual class is there instead of if it exists in a string
* Instead of just removing `closest` all together, check to see if the functionality exists. If it does exist, use it, otherwise fall back to the while loop. It's a little more code, but it means that for browsers that support that emerging standard, they'll be able to use a native method (which will be faster)

Snugug commented:
Yah, that's a [known compatibility issue](https://kangax.github.io/compat-table/es6/#test-const_basic_support_(strict_mode)) with Safari that should be resolved in the next version of Safari

Snugug commented:
![Pokémon Cheering](https://media.giphy.com/media/MhHXeM4SpKrpC/giphy.gif)
Celebrate!

ayush2k commented:
![Alt Text](https://d.gr-assets.com/hostedimages/1390790566ra/8278443.gif)

@Snugug thanks!
